hail/src/main/scala/is/hail/expr/ir/Emit.scala

```diff
@@ -1996,21 +2046,139 @@ private class Emit(
             )
           }
         }
+      case NDArrayMap2(lChild, rChild, lName, rName, body) =>
+        val lP = lChild.pType.asInstanceOf[PNDArray]
+        val rP = rChild.pType.asInstanceOf[PNDArray]
+
+        val lElemType = lP.elementType
+        val rElemType = rP.elementType
+
+        val lVti = typeToTypeInfo(lElemType.virtualType)
+        val rVti = typeToTypeInfo(rElemType.virtualType)
+
+        val lElemRef = coerce[Any](mb.newField(lName)(lVti))
+        val rElemRef = coerce[Any](mb.newField(rName)(rVti))
+
+        val bodyEnv = env.bind(name=lName, v=(lVti, false, lElemRef.load()))
+                         .bind(name=rName, v=(rVti, false, rElemRef.load()))
+        val bodyt = this.emit(body, bodyEnv, er, None)
+
+        val leftChildEmitter = deforest(lChild)
+        val rightChildEmitter = deforest(rChild)
+
+        val (unifiedShapeSetup, unifiedShape, unifiedShapePType) = NDArrayEmitter.unifyShapes(mb, leftChildEmitter.outputShape, leftChildEmitter.outputShapePType,
+          rightChildEmitter.outputShape, rightChildEmitter.outputShapePType, region)
+
+        val setup = Code(leftChildEmitter.setup, rightChildEmitter.setup, unifiedShapeSetup)
+
+        new NDArrayEmitter(mb, unifiedShapePType.size, unifiedShape,
+          lP.shape.pType, body.pType, setup) {
+          override def outputElement(idxVars: Array[Code[Long]]): Code[_] = {
+
+            val lIdxVars2 = NDArrayEmitter.zeroBroadcastedDims2(mb, idxVars, nDims, leftChildEmitter.outputShape, leftChildEmitter.outputShapePType)
+            val rIdxVars2 = NDArrayEmitter.zeroBroadcastedDims2(mb, idxVars, nDims, rightChildEmitter.outputShape, rightChildEmitter.outputShapePType)
+
+            Code(
+              lElemRef := leftChildEmitter.outputElement(lIdxVars2),
+              rElemRef := rightChildEmitter.outputElement(rIdxVars2),
+              bodyt.setup,
+              bodyt.m.orEmpty(Code._fatal("NDArray map body cannot be missing")),
```

Comment:
We can't lay down the missingness bit and do nothing about it!

`EmitArrayTriplet` should probably not take `Option[Code[Boolean]]` but some `Missingness` type that has named alternatives. For now, remember that `None` means definitely not-missing (see `MakeArray`). If the missingness is `None`, everything is fine, we should not throw an error. If the missing ness is `Some(m)` we need to emit `m.mux(Code._fatal(...), Code._empty)`.

Reply:
I don't use `EmitArrayTriplet` at all, a little confused by this comment.

Reply:
Ah, disregard. I was misreading this code, now I understand what orEmpty does.